packages/react/src/rise.tsx

```diff
@@ -57,10 +58,16 @@ export type ActionModelState<
   name: T
 } & K
 export type ActionModelStatePayload<T> = Omit<T, 'name' | '$'>
-export type ActionDefinition<T extends ActionModelState<any, any>> = {
-  action: (payload: ActionModelStatePayload<T>) => void
-  validate?: (args: unknown) => ActionModelStatePayload<T>
-}
+export type ActionsDefinition<ActionDefinition> = UnionToIntersection<
```

Comment:
before: I understood this type
after: I don't understand

not that it is wrong. just saying its hard to understand

Reply:
should we provide some documentation for it?

I agree it's complex. Not a huge fan of it. But it unlocks quite a bit of safety!

Let's suppose you have:
```
typeof action('navigate', { path }) // NavigateAction
typeof action('goBack') // GoBackAction
```
they are both ActionModelState. First one is ActionModelState<'navigate', { path: string }>, and second one is  ActionModelState<'goBack'>

When you create an object that is supposed to handle all different types of actions, you don't want to write:
```
const myActions = {
  'navigate': ActionPayload<NavigateAction>
}
```
because you're defining 'navigate' here and in case it changes in the other file where action helper is defined, you will have an unhandled action.

That's why I came up with this "helper" "ActionsDefinition" that takes union of all your actions.

You write ActionsDefinition<NavigateAction | GoBackAction> and it will do the following:

- When you pass union type to generic, it's basically "for each item in union type, apply this generic". So, ` ActionDefinition extends ActionModelState<infer Name, infer Action>` allows us to "unpack" Name and Action from each "ActionModelState" item in an union.
- From that point onwards, we just create a type that has "action" handler and "validate"

At this point `ActionsDefinition<NavigateAction | GoBackAction>` is converted into:

```
{ 'navigate': { action: (props) => void }} | {'goBack': { action: () => void }}
```

After wrapping everything with `UnionToIntersection`, we end up with a single object.

If there's
- argument mismatch
- wrong action name
- some actions are not handled
- extraneous actions are handled (no helper defined for such action or action was removed)
- payload type is different
- validator returns different type

You will receive type error.

Reply:
We can also return back to previous version, as it's more simple. Only challenge will be less type safety, but part of me feels like if we over engineer the types they will be holding us back really soon.

Reply:
After reading my own explanation I realised that while added type safety is nice, I don't think that complexity is worth it.